src/GraphCoarsen.cpp

```diff
@@ -13,788 +13,1100 @@
+
+    std::vector<int> row_map(local_size);
+
+    for (int i = 0; i < local_size; ++i)
+    {
+        assert(global_marker[local_indices[i]] >= 0);
+        row_map[i] = global_marker[local_indices[i]];
+    }
+
+    ClearMarker(global_marker, ext_indices);
+
+    return ext_mat.GetRow(row_map);
+}
+
+std::vector<int> GraphCoarsen::GetExtDofs(const ParMatrix& mat_ext, int row) const
+{
+    const auto& diag = mat_ext.GetDiag();
+    const auto& offd = mat_ext.GetOffd();
+
+    auto diag_dofs = diag.GetIndices(row);
+    auto offd_dofs = offd.GetIndices(row);
+
+    int diag_size = diag.Cols();
+
+    for (auto i : offd_dofs)
+    {
+        diag_dofs.push_back(i + diag_size);
     }
-    int cdof_counter2 = cdof_counter;
-    for (int i = cdof_counter2; i < ncdofs; i++)
+
+    return diag_dofs;
+}
+
+ParMatrix GraphCoarsen::MakeExtPermutation(const ParMatrix& parmat) const
+{
+    MPI_Comm comm = parmat.GetComm();
+
+    const auto& diag = parmat.GetDiag();
+    const auto& offd = parmat.GetOffd();
+    const auto& colmap = parmat.GetColMap();
+
+    int num_diag = diag.Cols();
+    int num_offd = offd.Cols();
+    int num_ext = num_diag + num_offd;
+
+    const auto& mat_starts = parmat.GetColStarts();
+    auto ext_starts = parlinalgcpp::GenerateOffsets(comm, num_ext);
+
+    SparseMatrix perm_diag = SparseIdentity(num_ext, num_diag);
+    SparseMatrix perm_offd = SparseIdentity(num_ext, num_offd, num_diag);
+
+    return ParMatrix(comm, ext_starts, mat_starts, std::move(perm_diag), std::move(perm_offd), colmap);
+}
+
+MixedMatrix GraphCoarsen::Coarsen(const MixedMatrix& mgl) const
+{
+    SparseMatrix agg_cdof_edge = BuildAggCDofEdge();
+    auto M_elem = BuildElemM(mgl, agg_cdof_edge);
+
+    SparseMatrix D_c = BuildCoarseD();
+    SparseMatrix W_c;
+
+    if (mgl.LocalW().Rows() == P_vertex_.Rows())
     {
-        select_i[cdof_counter + 1] = select_i[cdof_counter] + 1;
-        cdof_counter++;
+        SparseMatrix P_vertex_T = P_vertex_.Transpose();
+        W_c = P_vertex_T.Mult(mgl.LocalW().Mult(P_vertex_));
     }
-    assert(cdof_counter == ncdofs);
-    int* select_j = new int[ntruecdofs];
-    std::iota(select_j, select_j + ntruecdofs, 0);
-    mfem::SparseMatrix select(select_i, select_j, d_td_d_diag_data.GetData(), ncdofs,
-                              ntruecdofs, true, false, false);
-    mfem::HypreParMatrix select_d(comm, edge_cd_start_.Last(),
-                                  edge_ctd_start_.Last(), edge_cd_start_,
-                                  edge_ctd_start_, &select);
-    mfem::HypreParMatrix* d_td = ParMult(&d_td_d, &select_d);
-
-    return unique_ptr<mfem::HypreParMatrix>(d_td);
+
+    ParMatrix edge_true_edge = BuildEdgeTrueEdge();
+
+    return MixedMatrix(std::move(M_elem), std::move(agg_cdof_edge),
+                       std::move(D_c), std::move(W_c),
+                       std::move(edge_true_edge),
+                       BuildAggCDofVertex(),
+                       face_cdof_);
```

Comment:
Is it right that the last two matrices are copied in the constructor of `MixedMatrix`? For `face_cdof_`, I guess the reason is because it is a member in `GraphCoarsen` and there is a function to reference it. If we decided to not keep it in `GraphCoarsen`, can we also std::move it?

Reply:
Only `face_cdof_` is copied since it is a member, the `BuildAggCDofVertex` returns an rvalue and is moved into the constructor.  It should be possible to move `face_cdof_` too since it is not used after coarsening.